### DISCUSSION TITLE
Implémenter le formattage de nombre dans publicodes

### DISCUSSION THREAD

**Discussion by johangirod** [publicodes/publicodes#403] [Ideas]
## Existant

Aujourd'hui l'évaluation d'une expression numérique publicodes retourne un objet avec une clé `nodeValue` contenant un nombre javascript (`Number`), ainsi qu'une clé `unit` contenant l'unité. Pour afficher cette valeur formatée, on exporte une fonction [`formatValue`](https://publi.codes/docs/api#function-formatvalueevaluatednode-options) qui prend en entrée l'objet retourné `evaluate` et retourne une chaine de caractère.

## Limitations

1. Cette fonction est utilisée par défaut pour le formattage des valeur dans le mécanisme d'interpolation de chaîne de charactè-re `texte`. On souhaiterait pourtant pouvoir avoir plus de contrôle sur le formattage : par exemple pouvoir définir le nombre de décimales, ou utiliser les valeurs brutes sous forme "javascript".
2. Cela fait une fonction en plus exposée par l'API sans justification précise
3. Dans l'idée de créer une API REST depuis des règles publicodes, on voudrait pouvoir utiliser la version formattée d'une évaluation publicodes sans avoir à passer par un appel supplémentaire
4. La logique de [sérialisation d'évaluation](https://github.com/betagouv/publicodes/blob/master/packages/core/source/serializeEvaluation.ts) est proche du formattage, mais elle est dans une fonction séparée aujourd'hui.

## Propositions

Le formatage des valeurs pourrait avoir lieu directement en publicodes. Via un nouveau mécanisme chainable `format`. 
On peut imaginer les formes suivantes :

- `format:JSON` utilise la valeur JSON pour le formattage, c'est la même que la valeur affichée par console.log(evaluation.nodeValue)
- `format: publicodes` : c'est une valeur publicodes, elle peut être utilisée comme argument de setSituation, ou retravaillée dans une nouvelle expression publicodes.
-  `format: { unité , nombre chiffres, langues, ...options}`: un formatage pour être lu par les humains avec différentes options qui serait analogue à la fonction `formatValue`

**Comment by mquandalle** [comment]
voir aussi ce TODO dans notre code https://github.com/betagouv/mon-entreprise/blob/78af0c8734fb0a9ea13a846cc0c6ffdb68f191f3/site/source/components/conversation/NumberInput.tsx#L108
